### PR TITLE
Rewrite the html parser for (meta)infoboxes

### LINKED ISSUE #24: Attributes in meta infoboxes shouldn't map to the empty string
Some attributes map to an empty string (`u''`) when fetched using:

```python
from wikipediabase.infobox import get_meta_infobox
get_meta_infobox('<TEMPLATE_NAME>').attributes
```

Any idea why this is happening?

Below is a list of tuples in the form `(attribute, [list of infoboxes for which this happens])` for all attributes that map to the empty string:
```
(u'DST_note', [u'Template:Infobox country'])
(u'nebty_hiero', [u'Template:Infobox pharaoh'])
(u'junction', [u'Template:Infobox road'])
(u'female_weight', [u'Template:Infobox animal breed'])
(u'mottotranslated', [u'Template:Infobox law enforcement agency'])
(u'life_span', [u'Template:Infobox former subdivision'])
(u'coach_year', [u'Template:Infobox college inline hockey team', u'Template:Infobox college ice hockey team'])
(u'year_end', [u'Template:Infobox former country'])
(u'motto_translated', [u'Template:Infobox law enforcement agency'])
(u'female_height', [u'Template:Infobox animal breed'])
(u'birth_place', [u'Template:Infobox family'])
(u'femaleheight', [u'Template:Infobox goat breed', u'Template:Infobox donkey', u'Template:Infobox pig breed', u'Template:Infobox water buffalo breed', u'Template:Infobox cattle breed', u'Template:Infobox sheep breed'])
(u'coach_ties', [u'Template:Infobox college inline hockey team', u'Template:Infobox college ice hockey team'])
(u'golden_hiero', [u'Template:Infobox pharaoh'])
(u'coach_losses', [u'Template:Infobox college inline hockey team', u'Template:Infobox college ice hockey team'])
(u'year_start', [u'Template:Infobox former country'])
(u'area_note', [u'Template:Infobox settlement'])
(u'population_note', [u'Template:Infobox settlement'])
(u'femaleweight', [u'Template:Infobox goat breed', u'Template:Infobox poultry breed', u'Template:Infobox donkey', u'Template:Infobox pig breed', u'Template:Infobox water buffalo breed', u'Template:Infobox cattle breed', u'Template:Infobox sheep breed'])
(u'coach_wins', [u'Template:Infobox college inline hockey team', u'Template:Infobox college ice hockey team'])
```

### PR DESCRIPTION
**Pull request by cordeliaavery**
Fixes #24

### CODE REVIEW

**Comment by cordeliaavery** [comment]
still needs some additional test cases, and doesn't solve the parsing issue in issue #95

File: wikipediabase/infobox.py [hunk 1]
```diff
@@ -739,51 +741,53 @@ def parse_infobox_html(html_source):
...
+                    vals = re.findall(r'!!!!!([^!]+)!!!!!', t.get_text())
+                    for x in range(min(len(keys), len(vals))):
+                        key = clean_key(keys[x])
+                        value = clean_value(vals[x])
+
+                        if key and value:
+                            if not re.search('!!!!!([^!]+)!!!!!', value):
+                                # The regex eliminates the '!!!!!' for some reason,
+                                # and without it this won't be added as an attribute
+                                value = '!!!!!' + value + '!!!!!'
```
Comment (alvaromorales) [inline wikipediabase/infobox.py]: Nice catch! You could also change the regex:

```python
if not re.search(r'(!!!!![^!]+!!!!!)', value):
    pass
```

File: wikipediabase/util.py [hunk 2]
```diff
@@ -216,3 +216,51 @@ def output(s):
     if not isinstance(s, unicode):
         raise StringException("Not a unicode string: %s" % s)
     return s
+
```
Comment (alvaromorales) [inline wikipediabase/util.py]: moving these functions to `util.py` makes a lot of sense

File: wikipediabase/util.py [hunk 3]
```diff
@@ -216,3 +216,51 @@ def output(s):
...
+    # be a colon after the attribute name (see https://en.wikipedia.org/wiki/Template:Taxobox
+    # -- expects attribute 'Family:' not 'Family')
+    # Will probably want to update this at some point, and can include the following filter:
+
+    # key = re.sub(r':', '', key)
+
+    return key.strip()
+
+
+def clean_value(e_val):
```
Comment (alvaromorales) [inline wikipediabase/util.py]: should we remove nbsp (unicode 160) in this function?

**Comment by alvaromorales** [comment]
Great work!

> still needs some additional test cases

What kinds of test cases? Are you planning on adding them to this PR?

Also, is this change supposed to remove `\u2022` in all rendered values? Cause I found some bullet points in:

```python
In [2]: get_meta_infobox('Template:Infobox settlement').attributes
{u'area-blank1-acre': u'\u2022 {{{area_blank1_title}}}',
 u'area-blank1-ha': u'\u2022 {{{area_blank1_title}}}',
 u'area-blank1-sq-mi': u'\u2022 {{{area_blank1_title}}}',
[...]
```

Also, I noticed some attributes are still being mapped to the empty string.

```python
In [3]: get_meta_infobox(u'Template:Infobox animal breed').attributes
Out[3]:
{u'alt-name': u'Other names',
[...]
 u'female-height': u'',
 u'female-weight': u'',
[...]
 u'wool-color': u'Wool color'}
```

When this happens, can we just exclude attributes that map to the empty string?

File: wikipediabase/infobox.py (hunk 1 above)
Comment (cordeliaavery) [inline wikipediabase/infobox.py]: @alvaromorales including these values actually gives us more information than we were able to extract previously (the changes to tests/test_knowledgebase.py aren't necessary if we just skip this regex)

**Comment by cordeliaavery** [comment]
I specifically excluded key values that were only a single bullet point. There were a couple of instances where we only had a bullet point and rendered value, and no corresponding markup value, so i excluded these; i hadn't considered cases where there were still bullet points in the values. I can pull them out, if you think we should.

Also the mapping to the empty string is really strange--when I look at the attributes from 'Template:Infobox animal breed' I don't get anything mapped to the empty string. I check both the keys and values when parsing the html, and don't add them unless they are both non-empty.

**Comment by alvaromorales** [comment]
lgtm! :ship:

**Pull request merged**